### PR TITLE
Lex unboxed float and int literals as single lexemes

### PR DESCRIPTION
**Pull request by ncik-roberts**
**Reviewer:** @goldfirere 

Lex these, and things like them, as single lexemes:
```
#2
#2L
#2.0
```

This fixes an undesirable parse failure on `f #2.0 #4.0`. There are alternatives in parsing, but in discussion with @goldfirere we decided the lexing-based approach was much simpler.

### Directives

We thought we couldn't treat an unboxed int, like `#1093`, as a single lexeme because of directives:  
```
#1093 "filename.ml"
```
The lexer produces directives. We'd expect it to produce a directive here instead of an unboxed int followed by a string.

This ends up being unproblematic:
* This PR makes the lexer treat things that look like directives as directives, and everything else as unboxed ints. It just requires more code in the lexer.
* `#1093` isn't a valid unboxed int, anyway, as an unboxed int requires a suffix.

Regardless, the potential ambiguity with directives is a property of the unboxed int *syntax*, not the lexing-vs.-parsing approach, so this PR doesn't make the world worse.

### Implementation

We now call the `directive` lexer in two places: one where we've already consumed the ensuing integer literal, and one where we haven't (the existing place). I considered either adding a *new* lexer for "directive after the first integer literal", but I found it easier to call the existing lexer in both places, have the lexer's regex be able to handle both circumstances, and do validity checks within the semantic action of the lexer.

### CODE REVIEW

File: ocaml/parsing/lexer.mll [hunk 1]
```diff
@@ -384,6 +410,16 @@ let add_docstring_comment ds =
 
 let comments () = List.rev !comment_list
 
+let float ~maybe_hash lit modifier =
+  match maybe_hash with
+  | "#" -> HASH_FLOAT (lit, modifier)
+  | _ -> FLOAT (lit, modifier)
```
Comment (goldfirere) [inline ocaml/parsing/lexer.mll:416]: This case presumably will always be the empty string? I think it's better to be explicit about that and then have a catch-all case that witnesses an error in the lexer.

**Review by goldfirere — APPROVED** [review]
Looks great. Thanks.

**Comment by goldfirere** [comment]
Not merging because this is targeting a branch of yours -- so merge when you feel it is appropriate.

File: ocaml/parsing/lexer.mll (hunk 1 above)
Comment (ncik-roberts) [reply ocaml/parsing/lexer.mll]: Sounds good to me.

**Pull request merged**